### PR TITLE
Add structure viewer to the analysis report 

### PR DESCRIPTION
**Pull request by SarahAlidoost**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [ ] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] ~~Your PR is about CNS~~
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [x] Your code follows our coding style
- [ ] ~~You wrote tests for the new code~~
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] ~~Your PR is about writing documentation for already existing code :fire:~~
- [ ] ~~Your PR is about writing tests for already existing code :godmode:~~

---

This PR adds the [NGL](http://nglviewer.org/ngl/gallery/) molecule viewer to the `report.html` file generated by `analyse` command. It also refactors tables in the report and now uses some HTML and CSS syntax. :see_no_evil:
Here is the list of further improvements which can be developed in other PRs: 
- enbale toggling clusters between plots and tables
- enable downloading all pdb files, and plots in one-go
- enable select 4 best structures
- be able to sort inside the table using cluster Rank
- add some of the ngl tools e.g. rotate, show water

### CODE REVIEW

**Comment by SarahAlidoost** [comment]
@mgiulini and @sverhoeven  can you please review this PR? Thanks.

**Comment by SarahAlidoost** [comment]
need to be fixed:
- NGL bg color --> white
- use cluster rank instead of id --> tables and plots

File: src/haddock/libs/libplots.py
Comment (sverhoeven) [inline src/haddock/libs/libplots.py:663]: I think this code does nothing. As we are loading files from the local http server. Can it be removed?

```suggestion
```

File: src/haddock/libs/libplots.py
Comment (sverhoeven) [inline src/haddock/libs/libplots.py:847]: When more scores are added then you can no longer see the best structures. You will need to scroll down to see them, which a user might not know.

Can the height be made dependent on the number of rows? For example use `12rem` as their are 1 header + 11 score rows + 1 best structure row.

File: src/haddock/libs/libplots.py
Comment (sverhoeven) [inline src/haddock/libs/libplots.py:726]: The p-tag adds margin above and below making it look different then the other cells. 

Can you use `span` tag instead so each cell has same height?

**Review by sverhoeven — APPROVED** [review]
Works for me. I am able to download and view the structures.

I have some inline suggestions which are mainly on cosmetics. So could also be done later.

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py]: Done!

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py:693]: now, the viewer shows the file name with the download link.

File: src/haddock/clis/cli_analyse.py
Comment (SarahAlidoost) [reply src/haddock/clis/cli_analyse.py]: added!

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py]: Do you mean no vertical scroll in the table? then, I set the `height=auto`.  The table will automatically adjust its height based on the content. Is this what you meant? :thinking:

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py]: fixed!

**Comment by SarahAlidoost** [comment]
@mgiulini could you please review this PR?

**Comment by SarahAlidoost** [comment]
> Works for me. I am able to download and view the structures.
> 
> I have some inline suggestions which are mainly on cosmetics. So could also be done later.

Thanks for reviewing and good comments :+1:

**Comment by mgiulini** [comment]
> @mgiulini could you please review this PR?

sure, I will do it first thing tomorrow!

File: src/haddock/clis/cli_analyse.py
Comment (mgiulini) [inline src/haddock/clis/cli_analyse.py:305]: what is this change? probably related to some space..try to avoid introducing such changes, as they make the history messy

File: src/haddock/libs/libplots.py
Comment (mgiulini) [inline src/haddock/libs/libplots.py:954]: could you add a small docstring to this function? explaining what is figures (array of plots, I suppose)

File: src/haddock/libs/libplots.py
Comment (mgiulini) [inline src/haddock/libs/libplots.py:861]: can you add a small explanation/docstring?

File: src/haddock/libs/libplots.py
Comment (mgiulini) [inline src/haddock/libs/libplots.py:922]: docstring (see above)

File: src/haddock/libs/libplots.py
Comment (mgiulini) [inline src/haddock/libs/libplots.py:930]: docstring missing (see above)

File: src/haddock/libs/libplots.py
Comment (mgiulini) [inline src/haddock/libs/libplots.py:757]: very minor, but why is this renumbering necessary? why not keeping the standard `cluster_N_model_M.pdb` nomenclature? I don't like seeing zeros in the file names eheh

**Review by mgiulini — APPROVED** [review]
nice PR! it works nicely on a set of runs I analysed.

I just added a couple of comments about the introduction of docstrings and naming convention.

**Comment by mgiulini** [comment]
I noticed a serious problem with the PR: with the previous implementation, the bar plots were built so as each cluster until rank 10, and subsequent clusters were lumped together in a single bar (the eleventh bar). As an example consider this plot
image


Now the clusters with rank > 10 are superimposed on the eleventh bar, making the plot quite ugly:
image

can you please check?

**Comment by SarahAlidoost** [comment]
> I noticed a serious problem with the PR: with the previous implementation, the bar plots were built so as each cluster is shown until rank 10, and subsequent clusters were lumped together in a single bar (the eleventh bar). As an example consider this plot image
> 
> Now the clusters with rank > 10 are superimposed on the eleventh bar, making the plot quite ugly: image
> 
> can you please check?

Sure, I will check it. Can you tell me which example did you use to generate the plots?

**Comment by rvhonorato** [comment]
This is a great addition, do we still need anything before this is merged?

**Comment by SarahAlidoost** [comment]
> This is a great addition, do we still need anything before this is merged?

thanks, yes, I want to address the comments by Marco before merging it :blush:

**Comment by mgiulini** [comment]
> Sure, I will check it. Can you tell me which example did you use to generate the plots?

Sure, I used a modified version of the docking-protein-protein-test.cfg in examples/docking-protein-protein:

```
# ====================================================================
# Protein-protein docking example with NMR-derived ambiguous interaction restraints

# directory in which the scoring will be done
run_dir = "run1-test"
ncores = 40

# postprocess caprieval folders with haddock3-analyse
postprocess = true
mode = "local"
# CNS executable path - optional
# cns_exec = "path/to/bin/cns"

# molecules to be docked
molecules =  [
    "data/e2aP_1F3G.pdb",
    "data/hpr_ensemble.pdb"
    ]

# ====================================================================
# Parameters for each stage are defined below, prefer full paths
# ====================================================================
[topoaa]
#tolerance = 20
autohis = false
[topoaa.mol1]
nhisd = 0
nhise = 1
hise_1 = 75
[topoaa.mol2]
nhisd = 1
hisd_1 = 76
nhise = 1
hise_1 = 15

[rigidbody]
tolerance = 20
ambig_fname = "data/e2a-hpr_air.tbl"
sampling = 200

[caprieval]
#tolerance = 20
reference_fname = "data/e2a-hpr_1GGR.pdb"

[clustfcc]

[caprieval]
#tolerance = 20
reference_fname = "data/e2a-hpr_1GGR.pdb"

[seletopclusts]

[caprieval]
#tolerance = 20
reference_fname = "data/e2a-hpr_1GGR.pdb"
```

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py]: added!

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py:863]: added!

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py:956]: added!

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py]: fixed!

File: src/haddock/libs/libplots.py
Comment (SarahAlidoost) [reply src/haddock/libs/libplots.py:931]: added!

**Comment by SarahAlidoost** [comment]
@mgiulini thanks for checking this thoroughly. The problem in the box plot is fixed now. In addition to this fix, there are some new changes:
-  In this pull request, the legends of the plots (both scatters and box plots) are now cluster ranks instead of cluster id, as discussed in the previous meeting. I noticed that the legend is not sorted. The order is also fixed now. 
- I changed the `px_colors.qualitative.Alphabet` to `px_colors.qualitative.Dark24`. The reason is that `qualitative.Alphabet` has a very light color that might be picked by plots, see number 8 [here](https://plotly.com/python/discrete-color/).

**Comment by SarahAlidoost** [comment]
@mgiulini and @rvhonorato I think the failing test is not related to the changes in this PR. Otherwise, please let me know.

**Comment by rvhonorato** [comment]
Its failing because of some CI oversight, addressed it in #491 - good to merge if you greenlight @mgiulini

**Comment by mgiulini** [comment]
hey @SarahAlidoost , I checked the new code, it works perfectly, well done! I'll merge it!

**Pull request merged**
